packages/flutter/lib/src/widgets/spell_check.dart

```diff
@@ -0,0 +1,322 @@
+    final String text = value.text;
+    final TextRange composingRegion = value.composing;
+    final TextStyle composingTextStyle =
+        style?.merge(const TextStyle(decoration: TextDecoration.underline)) ??
+            const TextStyle(decoration: TextDecoration.underline);
+    final TextStyle misspelledJointStyle =
+        style?.merge(misspelledStyle) ?? misspelledStyle;
+    bool textPointerWithinComposingRegion = false;
+    bool currSpanIsComposingRegion = false;
+
+    while (spellCheckSuggestions != null &&
+        textPointer < text.length &&
+        currSpanPointer < spellCheckSuggestions.length) {
+      currSpan = spellCheckSuggestions[currSpanPointer];
+
+      if (currSpan.range.start > textPointer) {
+        endIndex = currSpan.range.start < text.length
+            ? currSpan.range.start
+            : text.length;
+        textPointerWithinComposingRegion =
+            composingRegion.start >= textPointer &&
+                composingRegion.end <= endIndex &&
+                !composingWithinCurrentTextRange;
+
+        if (textPointerWithinComposingRegion) {
+          _addComposingRegionTextSpans(tsTreeChildren, text, textPointer,
+              composingRegion, style, composingTextStyle);
+          tsTreeChildren.add(
+            TextSpan(
+              style: style,
+              text: text.substring(composingRegion.end, endIndex)
+            )
+          );
+        } else {
+          tsTreeChildren.add(
+            TextSpan(
+              style: style,
+              text: text.substring(textPointer, endIndex)
+            )
+          );
+        }
+
+        textPointer = endIndex;
+      } else {
+        endIndex =
+            currSpan.range.end < text.length ? currSpan.range.end : text.length;
+        currSpanIsComposingRegion = textPointer >= composingRegion.start &&
+            endIndex <= composingRegion.end &&
+            !composingWithinCurrentTextRange;
+        tsTreeChildren.add(
+          TextSpan(
+            style: currSpanIsComposingRegion
+                ? composingTextStyle
+                : misspelledJointStyle,
+            text: text.substring(currSpan.range.start, endIndex)
+          )
+        );
+
+        textPointer = endIndex;
+        currSpanPointer++;
+      }
+    }
+
+    if (textPointer < text.length) {
+      if (textPointer < composingRegion.start &&
+          !composingWithinCurrentTextRange) {
+        _addComposingRegionTextSpans(tsTreeChildren, text, textPointer,
+            composingRegion, style, composingTextStyle);
+
+        if (composingRegion.end != text.length) {
+          tsTreeChildren.add(
+            TextSpan(
+              style: style,
+              text: text.substring(composingRegion.end, text.length)
+            )
+          );
+        }
+      } else {
+        tsTreeChildren.add(
+          TextSpan(
+            style: style, text: text.substring(textPointer, text.length)
+          )
+        );
+      }
+    }
```

Comment:
Can you elaborate what this last part does? Maybe we can combine this logic into the above while loop?

Reply:
This adds any remaining text as `TextSpan`s to the tree after all misspelled words have been added to it since the loop above just runs until that point. I could change that to include this logic, but I don't think it would make the code any simpler assuming that's the motivation. Maybe I can just add a comment here?

Reply:
Yes, a comment would be good!